server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java

```diff
@@ -108,19 +123,16 @@ public LeafBucketCollector getLeafCollector(final LeafReaderContext ctx, final L
             return new LeafBucketCollectorBase(sub, null) {
                 @Override
                 public void collect(int parentDoc, long bucket) throws IOException {
-                    // if parentDoc is 0 then this means that this parent doesn't have child docs (b/c these appear always before the parent
-                    // doc), so we can skip:
-                    if (parentDoc == 0 || parentDocs == null || childDocs == null) {
+                    // parentDoc can be 0 when searching:
```

Comment:
@kkewwei just to clarify 

> // if parentDoc is 0 then this means that this parent doesn't have child docs (b/c these appear always before the parent

this is not valid comment?

Reply:
I need to dig into this. 

The parent/child block join in Lucene relies on the behavior of the old comment -- child documents are written immediately before their parent.

Reply:
I aggree that the child documents are written immediately before their parent. the `parentDoc` here is not the documnet parent, it should be more appropriately called the `searchParentDoc`. 

For example, nested2 can be `parentDoc`, but it is not the documnet parent.
```
POST index1_nest111/_search
{
  "aggregations": {
    "out_nested": {
      "aggregations": {
        "out_terms": {
          "aggregations": {
            "inner_nested": {
              "aggregations": {
                "inner_terms": {
                  "terms": {
                    "field": "nested1.name"
                  }
                }
              },
              "nested": {
                "path": "nested1"
              }
            }
          },
          "terms": {
            "field": "nested2.age"
          }
        }
      },
      "nested": {
        "path": "nested2"
      }
    }
  },
  "size": 0
}
```

Reply:
https://private-user-images.githubusercontent.com/16730433/324480985-7401b8a9-2648-4656-bc53-b730107fab24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNjAyNjgsIm5iZiI6MTcxNTE1OTk2OCwicGF0aCI6Ii8xNjczMDQzMy8zMjQ0ODA5ODUtNzQwMWI4YTktMjY0OC00NjU2LWJjNTMtYjczMDEwN2ZhYjI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDA5MTkyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyMjE3ZWFhNjhiZmM2M2ExMjdkMGY5OGYyYjg5OTZkMzYwOTkyYzUxMTk3NjJmNzY5ZDYxOWY5YWFjYTI5OGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NB-wYtVAodrBBO4_WO-mgCqiScWsOj4zacrNQyQqxBk

Reply:
@msfroh The issue is specifically for the nested NestedAggregator subagg (confusing!)

@kkewwei is creating 2 segments in his reproduction https://github.com/opensearch-project/OpenSearch/issues/13303#issuecomment-2069477578 and we are basically missing the first document in the sub-agg for each segment.

Reply:
Ahh... so the problem come down to a mismatch on the word "parent". 

In one case, it's talking about the parent doc (from the Lucene block-join standpoint), while in the other case, it's talking about a child doc ID (from the block-join standpoint) from the parent aggregation, where we're trying to aggregate over a sibling of that child. So, we need to map from that document to its parent, then join to the appropriate sibling. 

Is that a fair representation of the situation?

Reply:
Yes, that's what I mean, I should distinguish it from the name.